src/Language/PureScript/Kinds.hs

```diff
@@ -1,109 +1,164 @@
+{-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveTraversable #-}
 
 module Language.PureScript.Kinds where
 
 import Prelude.Compat
 
 import GHC.Generics (Generic)
 import Control.DeepSeq (NFData)
+import Data.Function (fix)
 import Data.Text (Text)
 import qualified Data.Text as T
-import Data.Aeson.BetterErrors (Parse, key, asText, asIntegral, nth, fromAesonParser, toAesonParser, throwCustomError)
+import Data.Aeson.BetterErrors (Parse, key, asText, asIntegral, nth, fromAesonParser, toAesonParser, throwCustomError, (<|>))
 import Data.Aeson ((.=))
 import qualified Data.Aeson as A
 
+import Language.PureScript.AST.SourcePos
 import Language.PureScript.Names
 import qualified Language.PureScript.Constants as C
 
+type SourceKind = Kind SourceAnn
+
 -- | The data type of kinds
-data Kind
+data Kind a
   -- | Unification variable of type Kind
-  = KUnknown Int
+  = KUnknown a Int
   -- | Kinds for labelled, unordered rows without duplicates
-  | Row Kind
+  | Row a (Kind a)
   -- | Function kinds
-  | FunKind Kind Kind
+  | FunKind a (Kind a) (Kind a)
   -- | A named kind
-  | NamedKind (Qualified (ProperName 'KindName))
-  deriving (Show, Eq, Ord, Generic)
+  | NamedKind a (Qualified (ProperName 'KindName))
+  deriving (Show, Generic, Functor, Foldable, Traversable)
+
+instance NFData a => NFData (Kind a)
+
+srcKUnknown :: Int -> SourceKind
+srcKUnknown = KUnknown NullSourceAnn
+
+srcRow :: SourceKind -> SourceKind
+srcRow = Row NullSourceAnn
 
-instance NFData Kind
+srcFunKind :: SourceKind -> SourceKind -> SourceKind
+srcFunKind = FunKind NullSourceAnn
+
+srcNamedKind :: Qualified (ProperName 'KindName) -> SourceKind
+srcNamedKind = NamedKind NullSourceAnn
 
 -- This is equivalent to the derived Aeson ToJSON instance, except that we
```

Comment:
This comment is no longer accurate, because the derived ToJSON instance would put annotations into the `contents` field, right?

Reply:
It would, and I can just add it to the contents, or derive them. I don't think the backwards compatibility these instances were trying to preserve are still relevant (we haven't supported Star and Symbol as is for a while).

Reply:
I agree that the parser code handling Star and Symbol is very unlikely to still be relevant (although perhaps best to deal with that separately). I'm just suggesting that we update the comment to make it more accurate.